waku/v2/protocol/waku_store/waku_store.nim

```diff
@@ -612,14 +612,23 @@ proc queryFromWithPaging*(w: WakuStore, query: HistoryQuery, peer: RemotePeerInf
   return ok(messageList)
 
 proc queryLoop(w: WakuStore, query: HistoryQuery, candidateList: seq[RemotePeerInfo]): Future[MessagesResult]  {.async, gcsafe.} = 
-  ## loops through the candidateList in order and sends the query to each until one of the query gets resolved successfully
-  ## returns the retrieved messages, or error if all the requests fail
-  for peer in candidateList.items: 
-    let successResult = await w.queryFromWithPaging(query, peer)
-    if successResult.isOk: return ok(successResult.value)
-
-  debug "failed to resolve the query"
-  return err("failed to resolve the query")
+  ## loops through the candidateList in order and sends the query to each
+  ## once all responses have been received, the retrieved messages are consolidated into one deduplicated list
+  ## if no messages have been retrieved, the returned future will resolve into a MessagesResult result holding an empty seq.
+  var futureList: seq[Future[MessagesResult]]
+  for peer in candidateList.items:
+    futureList.add(w.queryFromWithPaging(query, peer))
+  futureList = await allFinished(futureList) # TODO: is there a proc that returns a Future[seq[T]?
+  var messagesList: seq[WakuMessage]
+  # all futures have already been awaited, so the following loop runs without actually waiting
+  for fut in futureList:
+    let successResult = await fut
```

Comment:
Since we "know" that these futures have all completed, why not rather just `read` the value (with a sanity check for completeness too). I think this will improve readability.
For example:
```
if fut.completed():
  let successResult = fut.read()
```
or something similar should be safe to do.

Reply:
You could also encompass this logic into the `op proc` parameter of a `mapIt` function and directly map your `seq[Future[MessagesResult]]` to a `seq[MessagesResult]`. With a bit more logic, and concatenating the different responses, the original `seq` of futures could also be mapped to a `seq[WakuMessage]` using a single `mapIt` function.

Reply:
Thank you. I did not know about `read()`. I will apply that.